File: api/go/budgets.go
package handler

import (
	"net/http"
	
	"github.com/budget-buddy/api/lib"
)

// Handler handles budget CRUD operations
func Handler(w http.ResponseWriter, r *http.Request) {
	config := lib.Config{
		RequireAuth:    true,
		AllowedMethods: []string{"GET", "POST", "PUT", "DELETE"},
		EnableCORS:     true,
	}

	handler := lib.CreateHandler(budgetHandler, config)
	handler(w, r)
}

func budgetHandler(w http.ResponseWriter, r *http.Request) {
	user, ok := lib.GetUserFromContext(r)
	if !ok {
		lib.ErrorResponse(w, "Unauthorized", http.StatusUnauthorized, nil)
		return
	}

	switch r.Method {
	case "GET":
		handleGetBudgets(w, r, user)
	case "POST":
		handleCreateBudget(w, r, user)
	case "PUT":
		handleUpdateBudget(w, r, user)
	case "DELETE":
		handleDeleteBudget(w, r, user)
	default:
		lib.ErrorResponse(w, "Method not allowed", http.StatusMethodNotAllowed, nil)
	}
}

func handleGetBudgets(w http.ResponseWriter, r *http.Request, user *lib.User) {
	period := lib.GetQueryParam(r, "period", "monthly")

	// TODO: Query database
	budgets := []map[string]interface{}{
		{
			"id":              "budget-1",
			"user_id":         user.ID,
			"category":        "Groceries",
			"amount":          500.0,
			"period":          period,
			"alert_threshold": 80,
			"created_at":      "2024-01-01T00:00:00Z",
		},
	}

	lib.SuccessResponse(w, map[string]interface{}{
		"budgets": budgets,
	}, http.StatusOK)
}

func handleCreateBudget(w http.ResponseWriter, r *http.Request, user *lib.User) {
	var input lib.CreateBudgetInput
	if err := lib.ParseJSONBody(r, &input); err != nil {
		lib.ErrorResponse(w, "Invalid JSON body", http.StatusBadRequest, nil)
		return
	}

	// Validate input
	if input.Category == "" {
		lib.ErrorResponse(w, "Category is required", http.StatusBadRequest, nil)
		return
	}

	if input.Amount <= 0 {
		lib.ErrorResponse(w, "Amount must be positive", http.StatusBadRequest, nil)
		return
	}

	if input.Period != "weekly" && input.Period != "monthly" && input.Period != "yearly" {
		lib.ErrorResponse(w, "Period must be 'weekly', 'monthly', or 'yearly'", http.StatusBadRequest, nil)
		return
	}

	// TODO: Insert into database
	budget := map[string]interface{}{
		"id":              "budget-new",
		"user_id":         user.ID,
		"category":        input.Category,
		"amount":          input.Amount,
		"period":          input.Period,
		"alert_threshold": input.AlertThreshold,
		"created_at":      "2024-01-15T10:00:00Z",
	}

	lib.SuccessResponse(w, map[string]interface{}{
		"budget": budget,
	}, http.StatusCreated)
}

func handleUpdateBudget(w http.ResponseWriter, r *http.Request, user *lib.User) {
	id := lib.GetQueryParam(r, "id", "")
	if id == "" {
		lib.ErrorResponse(w, "Budget ID required", http.StatusBadRequest, nil)
		return
	}

	var input map[string]interface{}
	if err := lib.ParseJSONBody(r, &input); err != nil {
		lib.ErrorResponse(w, "Invalid JSON body", http.StatusBadRequest, nil)
		return
	}

	// TODO: Update in database
	budget := map[string]interface{}{
		"id":      id,
		"user_id": user.ID,
		"updated": true,
	}

	lib.SuccessResponse(w, map[string]interface{}{
		"budget": budget,
	}, http.StatusOK)
}

func handleDeleteBudget(w http.ResponseWriter, r *http.Request, user *lib.User) {
	id := lib.GetQueryParam(r, "id", "")
	if id == "" {
		lib.ErrorResponse(w, "Budget ID required", http.StatusBadRequest, nil)
		return
	}

	// TODO: Delete from database

	lib.SuccessResponse(w, map[string]interface{}{
		"message": "Budget deleted successfully",
		"id":      id,
	}, http.StatusOK)
}

File: api/budgets.ts
/**
 * Serverless Function: Budgets API
 * Handles budget management operations
 */

import { VercelRequest, VercelResponse } from '@vercel/node';
import {
  createServerlessHandler,
  AuthenticatedRequest,
  parseBody,
  successResponse,
  errorResponse,
  getSupabaseClient,
} from './_lib/serverless-helpers';

async function handler(req: AuthenticatedRequest, res: VercelResponse): Promise<void> {
  const supabase = getSupabaseClient(req);

  // GET: Fetch budgets
  if (req.method === 'GET') {
    const { searchParams } = new URL(req.url!, `http://${req.headers.host}`);
    const period = searchParams.get('period') || 'monthly';

    const { data, error } = await supabase
      .from('budgets')
      .select('*')
      .eq('user_id', req.user!.id)
      .eq('period', period)
      .order('created_at', { ascending: false });

    if (error) {
      errorResponse(res, 'Failed to fetch budgets', 500, error);
    }

    successResponse(res, { budgets: data });
  }

  // POST: Create budget
  if (req.method === 'POST') {
    const body = await parseBody(req);

    if (!body || !body.category || !body.amount || !body.period) {
      errorResponse(res, 'Missing required fields', 400);
    }

    const { data, error } = await supabase
      .from('budgets')
      .insert({
        user_id: req.user!.id,
        category: body.category,
        amount: body.amount,
        period: body.period,
        start_date: body.start_date || new Date().toISOString(),
        end_date: body.end_date,
        alert_threshold: body.alert_threshold || 80,
      })
      .select()
      .single();

    if (error) {
      errorResponse(res, 'Failed to create budget', 500, error);
    }

    successResponse(res, { budget: data }, 201);
  }

  // PUT: Update budget
  if (req.method === 'PUT') {
    const body = await parseBody(req);
    const { searchParams } = new URL(req.url!, `http://${req.headers.host}`);
    const id = searchParams.get('id');

    if (!id) {
      errorResponse(res, 'Budget ID required', 400);
    }

    const { data, error } = await supabase
      .from('budgets')
      .update(body)
      .eq('id', id)
      .eq('user_id', req.user!.id)
      .select()
      .single();

    if (error) {
      errorResponse(res, 'Failed to update budget', 500, error);
    }

    successResponse(res, { budget: data });
  }

  // DELETE: Delete budget
  if (req.method === 'DELETE') {
    const { searchParams } = new URL(req.url!, `http://${req.headers.host}`);
    const id = searchParams.get('id');

    if (!id) {
      errorResponse(res, 'Budget ID required', 400);
    }

    const { error } = await supabase
      .from('budgets')
      .delete()
      .eq('id', id)
      .eq('user_id', req.user!.id);

    if (error) {
      errorResponse(res, 'Failed to delete budget', 500, error);
    }

    successResponse(res, { message: 'Budget deleted' });
  }

  errorResponse(res, 'Method not allowed', 405);
}

export default createServerlessHandler(handler, {
  auth: true,
  cors: true,
});
